[WIP] Implementation of Inverse Wishart distribution

nonconvexopt opened the pull request:
On the process of #68050
Contiguous to previous contribution #68588 implemented `torch.distributions.wishart.Wishart` with @neerajprad.
cc @neerajprad

While working on #68588, @dccastro provided initiative implementation of Inverse Wishart distribution based on PyTorch: [https://github.com/biomedia-mira/deepscm/blob/master/deepscm/distributions/torch_wishart.py](https://github.com/biomedia-mira/deepscm/blob/master/deepscm/distributions/torch_wishart.py)
Thank you for the neat and intuitive work to reference.

TODO:
- [x] Implement all functions
- [ ] Unit Test
- [x] Documentation

Note) Removed 'Add transform from Wishart distribution to Inverse Wishart distribution' from TODO since we added manual implementation for Inverse Wishart distribution.

nonconvexopt committed:
Fix errors in concatinating shape in '.rsample'

nonconvexopt committed:
Added tests checking value shapes for Wishart

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Change argument orders in '.__init__()' to be more sensible

nonconvexopt committed:
Fix test in 'TestAgainstScipy'

nonconvexopt committed:
Fix bugs in resampling singular samples.

nonconvexopt committed:
Fix test for checking value shape in wishart distribution.

nonconvexopt committed:
Added test for Wishart following the practice of Multivariate Normal

nonconvexopt committed:
Revert vectorization in resampling process for singular samples

nonconvexopt committed:
Debug added tests.

nonconvexopt committed:
Debug mypy check.

nonconvexopt committed:
Force to give 'df' parameter at initialization

nonconvexopt committed:
Debug tests.

nonconvexopt committed:
Fix errors in test.

nonconvexopt committed:
Added warinings for singular sample risk.

nonconvexopt committed:
Vectorize resampling singular samples.

nonconvexopt committed:
Test better way to deal singular samples.

nonconvexopt committed:
Debug argument error for 'scipy.stats.wishart'

nonconvexopt committed:
Try better way to deal singular samples.

nonconvexopt committed:
Reduce batch shape in '.log_prob'

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Debugging tests for distribution

nonconvexopt committed:
Modfied '_check_sampler_sampler' function to consider high-dimensional tensor.

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'wishart_distribution' of github.com:nonconvexopt/pytorch into wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'wishart_distribution' of github.com:nonconvexopt/pytorch into wishart_distribution

nonconvexopt committed:
Improve '_check_sampler_sampler' to consider high-dimensional inputs

nonconvexopt committed:
Improve '_check_sampler_sampler' to consider high-dimensional inputs

nonconvexopt committed:
Fix typo in '_check_sampler_sampler'

nonconvexopt committed:
Fix shape of '.log_prob'

nonconvexopt committed:
Fix wrong test.

nonconvexopt committed:
Refactored '.rsample', '.log_prob' and '.entropy'

nonconvexopt committed:
Don't reduce the batch shape of '.log_prob' and '.entropy'

nonconvexopt committed:
Fix typo in '.rsample'

nonconvexopt committed:
Fix wrong test and make 'test_wishart_moments' more finer

nonconvexopt committed:
Fix wrong test and reduce the random matrix size of 'test_wishart_moments'

nonconvexopt committed:
Make 'test_wishart_moments' more delicate

nonconvexopt committed:
Reduce the random variable shape in 'test_wishart_moments'

nonconvexopt committed:
Made test 'test_wishart_moments' easier

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Add '_log_normalizer' function

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Fix CI errors.

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'wishart_distribution' of github.com:nonconvexopt/pytorch into wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Fix resampling procedure to handle singleton shape sample and all singular sample.

nonconvexopt committed:
Refactor resampling code.

nonconvexopt committed:
Clean the assignment of instance variable 'self._reduced_batch_dims'

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Update wishart.py

nonconvexopt committed:
Change '.logical_not()' to '~'.

nonconvexopt committed:
Keep '.rsample' to '.sample' until testing.

nonconvexopt committed:
Change '.rsample' to '.sample' temporary.

nonconvexopt committed:
Revert the change from '.rsample' to '.sample' and use '.rsample'.

nonconvexopt committed:
Refactor '.rsample' and apply the implementation of neerajprad

nonconvexopt committed:
Fix CI errors.

nonconvexopt committed:
Fix jit.trace errors.

nonconvexopt committed:
Revert change in '.rsample'

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Remove the redundant computation in '.rsample'

nonconvexopt committed:
Merge branch 'wishart_distribution' of github.com:nonconvexopt/pytorch into wishart_distribution

nonconvexopt committed:
Fix tensor indexing error in '.rsample'.

nonconvexopt committed:
Debugging CI errors.

nonconvexopt committed:
Revert change in '.rsample'

nonconvexopt committed:
Debugging JIT CI error.

nonconvexopt committed:
Fix typo

nonconvexopt committed:
Relax the optimization for JIT trace in '.rsample'

nonconvexopt committed:
Reflected all feedbacks from the review of neerajprad

nonconvexopt committed:
Fix CI errors.

nonconvexopt committed:
Fix CI error.

nonconvexopt committed:
Process float 'df' argument.

nonconvexopt committed:
Refactor '__init__' to pass mypy test.

nonconvexopt committed:
Add warning docstring for '.rsample'.

nonconvexopt committed:
Change the name of the parameter in '.rsample' to imply its usage.

nonconvexopt committed:
Revise docstring.

nonconvexopt committed:
Process float type for 'df' parameter.

nonconvexopt committed:
Clean code in '.__init__()'

nonconvexopt committed:
Add test for float 'df' values.

nonconvexopt committed:
Fix test checking float 'df' values.

nonconvexopt committed:
Fit PEP 8 for 'torch/distributions/wishart.py'

nonconvexopt committed:
Add class variable needed for inheriting 'torch.distributions.exp_family.ExponentialFamily'

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Initial structure of inverse wishart distribution class referred from 'torch.distributions.wishart.Wishart'

nonconvexopt committed:
Seperate branch to #70275

nonconvexopt committed:
Initial structure of inveser wishart distribution class.

nonconvexopt committed:
Merge branch 'pytorch:master' into wishart_distribution

nonconvexopt committed:
Add device and dtype argument to torch.arrange and reflect the feedback of vadimkantorov

nonconvexopt committed:
Merge branch 'wishart_distribution' of github.com:nonconvexopt/pytorch into wishart_distribution

nonconvexopt committed:
Merge branch 'wishart_distribution' into inverse_wishart_distribution

nonconvexopt committed:
Revise typos in base structure.

nonconvexopt committed:
Clean initial '__init__' method and add descriptions.

nonconvexopt committed:
Implement '.mean()'.

nonconvexopt committed:
Implement ".variance()".

nonconvexopt committed:
Implement '.log_prob()'.

nonconvexopt committed:
Revise class description.

nonconvexopt committed:
Implement '.entropy()'

nonconvexopt committed:
Implement '._natural_params()'

nonconvexopt committed:
Trivial Revision of '.entropy()'.

nonconvexopt committed:
Reduce redundant computation in '._natural_params()'

nonconvexopt committed:
Merge branch 'wishart_distribution' into inverse_wishart_distribution

nonconvexopt committed:
Fix '._natural_params()'.

nonconvexopt committed:
Implement '_log_normalizer()'.

nonconvexopt committed:
Add documentation.

nonconvexopt committed:
Implement '.rsample()'.

nonconvexopt commented:
I implemented `.rsample()` by just inverting diagonal matrix in `.rsample()` of Wishart distribution Bartlett sampling for the initial implementation.
But, I am actively looking for better mechanisms to sampling it. i.e '4.5 A New Way to Sample the IWP' in [Student-t Processes as Alternatives to Gaussian Processes](https://arxiv.org/abs/1402.4306).
Plz let me know if there is another suggestion.

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution

nonconvexopt committed:
Fix variable name error.

nonconvexopt committed:
Merge branch 'inverse_wishart_distribution' of github.com:nonconvexopt/pytorch into inverse_wishart_distribution

nonconvexopt committed:
Fix error in docs.

nonconvexopt committed:
Register inverse wishart ditribution to import.

nonconvexopt committed:
Fix typo.

nonconvexopt committed:
Merge branch 'master' into inverse_wishart_distribution

nonconvexopt committed:
Register tests for Inverse Wishart distribution.

nonconvexopt committed:
Merge branch 'inverse_wishart_distribution' of github.com:nonconvexopt/pytorch into inverse_wishart_distribution

nonconvexopt committed:
Fix typo.

nonconvexopt committed:
Add BADEXAMPLE for wishart and inverse wishart distribution.

nonconvexopt committed:
Register EXAMPLE for Wishart distribution.

nonconvexopt committed:
Add same tests in wishart distribution for inverse wishart distribution.

nonconvexopt committed:
Add remained tests in Wishart distribution which was not added.

nonconvexopt commented:
This force push was for editing wrong commit message.

nonconvexopt committed:
Remove ambiguity when brodcasting tensors in 'torch.distributions.wishart.Wishart'

nonconvexopt committed:
Revert modification in broadcasting.

nonconvexopt committed:
Fix broadcatst error in '.variance()'.

nonconvexopt committed:
Denote shape of 'p' and raise ValueError in '.mean()'

nonconvexopt committed:
Added ValueError same in 'torch.distributions.wishart.Wishart'.

nonconvexopt committed:
Use torch.expand instead of torch.view.

nonconvexopt committed:
Set tests only for cases '.mean' can be calculated.

nonconvexopt commented:
Sorry for halting the process. I am preparing a new development environment this week and it stopped me do formal codings. Hope I can get back to this PR soon.

nonconvexopt committed:
Increase 'df' values for EXAMPLE for 'test_mean'.

nonconvexopt committed:
Change '.expand' to '.view' to expand tensor at backward dimensions.

nonconvexopt committed:
Fix typo.

nonconvexopt committed:
Fix typo.

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution

nonconvexopt committed:
Improve readability of '_mvdigamma'.

nonconvexopt committed:
Merge branch 'inverse_wishart_distribution' of github.com:nonconvexopt/pytorch into inverse_wishart_distribution

nonconvexopt committed:
Fit the requirements of the value of 'df' for calculating '.mean'.

nonconvexopt committed:
Test cases only when mean of inverse wishart distribution can be calculated.

nonconvexopt committed:
Reduce unnecessary variables.

nonconvexopt committed:
Remove unnecessary variables and fix expand error.

nonconvexopt committed:
Change '.expand' to '.view' and fix expand error in Wishart dist.

nonconvexopt committed:
Revert 'df' value changes in the tests for Wishart distribution.

neerajprad commented on torch/distributions/inverse_wishart.py:
```diff
@@ -0,0 +1,290 @@
+        p = self._event_shape[-1]  # has singleton shape
+
+        # Implemented Sampling using Bartlett decomposition
+        noise = self._dist_chi2.rsample(sample_shape).sqrt().reciprocal().diag_embed(dim1=-2, dim2=-1)
```
Could you mention what are the changes w.r.t. the Wishart distribution implementation? e.g. reciprocal of Chi2 samples here. Any other? There's a lot of duplicated code between these two classes and we should think about how to restructure to keep redundancy to a minimum. If there are many small subtle variations, then this isn't important and we can keep them separate.

nonconvexopt replied:
Thank you for discovering good point to improve.
I seperated the implementation of Wishart and Inverse Wishart distribution before because of 2 reasons:
1) Some method or property like `.mean()` can be calculated in different conditions of Wishart and Inverse Wishart. 
2) I expected there is more sophisticated sampling algorithm for Inverse Wishart which cannot applied in Wishart.
But I think I have to admit that two implementation shares too much duplicated parts.

Those are my references about the latest sampling techniques for Inverse Wishart:
[1] [Student-t Processes as Alternatives to Gaussian Processes](https://arxiv.org/abs/1402.4306)
[2] [Scalable Bayesian dynamic covariance modeling with variational Wishart and inverse Wishart processes](https://arxiv.org/abs/1906.09360)
[3] [A variational approximate posterior for the deep Wishart process](https://arxiv.org/abs/2107.10125)
I hope that me and other users in the future to easily apply Variational Inference techniques for Matrix like in [3] using the official implementation in PyTorch.
I will be happy to restructure and find a way to unite two class in one implementation while fitting this requirements.

neerajprad commented:
Let me know if this is ready for a detailed review.

nonconvexopt commented:
> Let me know if this is ready for a detailed review.

I appreciate your kindness. I will look forward to request detailed review soon. I assume I can pass the remained CI tests until the weekend.

nonconvexopt committed:
Remove ambiguity of shape in tensor operation in '.mean()'

nonconvexopt committed:
Improve numerical stability and modifiy operation code style

nonconvexopt committed:
Add Valueerror checking condition for '.variance'

nonconvexopt committed:
Modify operation code style

nonconvexopt committed:
Fix typo

nonconvexopt committed:
Add warning for calculating variance for singular Inverse Wishart Distribution.

nonconvexopt committed:
Limit positive test cases to cases when variance can be calculated

nonconvexopt commented:
First two moments of Inverse Wishart distribution can be calculated in constrained cases differing whether it is Singular or non-Singular cases. Reference: (On the mean and variance of the generalized inverse of a singular Wishart matrix)[https://projecteuclid.org/journals/electronic-journal-of-statistics/volume-5/issue-none/On-the-mean-and-variance-of-the-generalized-inverse-of/10.1214/11-EJS602.full]

nonconvexopt committed:
Enhance bad examples

nonconvexopt committed:
Fix wrong positive example

nonconvexopt committed:
Fix typo in '.mean()'

nonconvexopt committed:
Limit test cases for Inverse Wishart to the cases when moments can be calculated

nonconvexopt committed:
Limit test cases

nonconvexopt committed:
Remove negative example which moments cannot be calculated

nonconvexopt committed:
Make tests for Inverse Wishart to focus on its own purpose

nonconvexopt commented:
I think it might be better to add positive semi-definite constraint to handle more general matrix distributions, Singular Wishart and Singular Inverse Wishart cases. Currently, only positive denfinite constraint is applied to the support and `.rsample()` triggers extensive resampling in Wishart/Inverse Wishart to fit the constraints. However, we may allow singular matrix as covariance or precision argument and also sampled matrices.

nonconvexopt commented:
Referring #71382, cholesky decomposition of Singular(positive-semidefinite) matrix is not available for now. I will complete full ranked Inverse Wishart in this PR for now. Hope I can apply the generalization to Sigular Wishart/Inverse Wishart in the future, too.

nonconvexopt committed:
Generalize to SIngular cases

nonconvexopt committed:
Focus on full rank case for now

This reverts commit f35a85f3a37e6cf401470ac50ec246a5bb7e3a3b.

nikitaved commented:
@nonconvexopt , as a follow-up on your LDL issue and the forward/backward AD support, I do not immediately see how to make it differentiable for singular outputs UNLESS the objective function does not depend on the kernel. The reason being is that the matrix D has to be invertible in order for dL to be reconstructable fully. Does your application involve a lot of singular inputs? I think we can work it out but with some work involving Sylvester equations or something like that...

nonconvexopt commented:
> @nonconvexopt , as a follow-up on your LDL issue and the forward/backward AD support, I do not immediately see how to make it differentiable for singular inputs UNLESS the objective function does not depend on the kernel (so that the corresponding dL parts are zeros). The reason being is that the matrix D has to be invertible in order for dL to be reconstructable fully. Does your application involve a lot of singular inputs? I think we can work it out but with some work involving Sylvester equations or something like that...

Thank you for letting me know. I heard Sylvester equations first time and it seems useful in linalg library. I hope I can work with you in that part to gain deeper understanding in PyTorch kernels written in C/C++. Is there anything I can work on together? 

> The reason being is that the matrix D has to be invertible in order for dL to be reconstructable fully
I don't understand this part. Even if there is zero diagonal elements in diagonal D, inverting only non-zero elements does not solve the problem?

nikitaved commented:
> @nikitaved Thank you for letting me know. I heard Sylvester equations first time and it seems useful in linalg library. I hope I can work with you in that part to gain deeper understanding in PyTorch kernels written in C/C++. Is there anything I can work on together?

Sure, we can work it out :) Also, feel free to ping me on reviews.


> > The reason being is that the matrix D has to be invertible in order for dL to be reconstructable fully
> 
> I don't understand this part. Even if there is zero diagonal elements in diagonal D, inverting only non-zero elements does not solve the problem?

This computes the Moore-Penrose pseudo-inverse of D, and it gives a projector onto the range of D, and the subspaces not in this range will be mapped to zero. But these subspaces carry useful information unless the final objective function does not depend on them.
I am currently working on a rank-restricted SVD decomposition that zeros out close to zero subspaces and ignores them in the backward step. Maybe we can provide such a function for the LDL decomposition as well. Just a note, singular inputs are not well-behaved across a multitude of linear algebra operations in PyTorch, be it forward (like solve) or backward(almost all of them involving any kind of factorization).

nonconvexopt committed:
Revert clipping for numerical stability

nonconvexopt committed:
Make positive tests to satisfy assertions for moment methods

nonconvexopt commented:
Thank you for your kind. I hope to get feedbacks from your review in the future. Recently, I tried using `torch.lobpcg` and I hoped SVD-like algorithms such as `torch.lobpcg` provides automatic cutoff based on eigen/singular value proportion offsets (like 0.95 or 0.99).
 
> I am currently working on a rank-restricted SVD decomposition that zeros out close to zero subspaces and ignores them in the backward step. Maybe we can provide such a function for the LDL decomposition as well.

This seems interesting topic for me, too. Can you introduce your PR?

nonconvexopt committed:
Fix typo in data type

nonconvexopt committed:
Remove bad examples with invalid variance temporary

nonconvexopt committed:
Fix typo in bad examples

nonconvexopt committed:
Remove BAD_EXAMPLE with invalid variances temporary

nonconvexopt committed:
Fiw wrong natural params

nonconvexopt committed:
Debug ExponentialFamily inheritance of Wishart together

nonconvexopt committed:
Extend exponential family to multivariate natural params

nonconvexopt committed:
Apply tests again

nonconvexopt committed:
Reduce df in some tests and make test harder

nonconvexopt committed:
Make tests to also consider numerical stability

nonconvexopt committed:
Make tests to also consider numerical stability

nonconvexopt committed:
Debug .log_prob()

nonconvexopt committed:
Debug .log_prob()

nonconvexopt committed:
Debug implementation and optimize .log_prob()

nonconvexopt committed:
Fix typo and fit flake8

nonconvexopt committed:
constrain support to positive-definite temporary

nonconvexopt committed:
Improve numerical stability of reciprocal

nonconvexopt committed:
Fix typo

nonconvexopt committed:
Refactor function usage

nonconvexopt committed:
Try using torch.script for internal functions

nonconvexopt committed:
Merge branch 'master' into inverse_wishart_distribution

nonconvexopt committed:
Clean codes

nonconvexopt committed:
Revert applying torch.script

nonconvexopt committed:
Improve numerical stability

nonconvexopt committed:
Check numerical stability more rigorously

nonconvexopt committed:
Seperate the changes in wishart to #72059

nonconvexopt committed:
Fix typo

nonconvexopt committed:
Set df values in test larger than (ndim - 1)

nonconvexopt committed:
Reduce ambiguity in the tests and debug df value error

nonconvexopt committed:
Fix typo

nonconvexopt committed:
Extend .entropy to consider multivariate natural parameters

nonconvexopt committed:
Increase tolerance value for .assertEqual

nonconvexopt committed:
Consider SciPy version and adjust test conditions

nonconvexopt committed:
Merge branch 'master' into distributions/wishart_improvement_v1

nonconvexopt committed:
Remove trailing spaces

nonconvexopt committed:
Fix typo and debug error in version comparison

nonconvexopt committed:
reduce variance in MC estimate

nonconvexopt committed:
Check correct scipy version allowing df < ndim for Wishart

nonconvexopt committed:
Allow higher error in MC estimate for moment

nonconvexopt committed:
Apply harder clamping

nonconvexopt committed:
Remove unnecessary code pieces

nonconvexopt committed:
Remove experimental changes

nonconvexopt committed:
Test numerical instability only in relevant tests

nonconvexopt committed:
Merge branch 'master' into distributions/wishart_improvement_v1

nonconvexopt committed:
Return correct natural paramters

nonconvexopt committed:
Extend exponential family to multivariate

nonconvexopt committed:
Debug _log_noramlizer

nonconvexopt committed:
Fit flake8

nonconvexopt committed:
Extend ExponentialFamily to multivariate

nonconvexopt committed:
Remove trailing spaces

nonconvexopt committed:
Merge branch 'master' into inverse_wishart_distribution

nonconvexopt committed:
Extend ExponentialFamily to multivariate

nonconvexopt commented:
Merged master and did something wrong in the `rebase`. Commits is `master` branch appears on this PR too and additional commits here will send a message to other PR. Searching solutions.

nonconvexopt commented:
This seems to be a good reference to recover `git rebase` mistake: [How to undo a mistaken git rebase (LIFE SAVER)](https://medium.com/@shreyawhiz/how-to-undo-a-mistaken-git-rebase-life-saver-2977ff0a0602)

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution

nonconvexopt committed:
Merge branch 'master' into inverse_wishart_distribution

nonconvexopt committed:
Merge branch 'master' of github.com:pytorch/pytorch into inverse_wishart_distribution

nonconvexopt committed:
fix singular input error in .log_prob

nonconvexopt committed:
remove assertion in .variance and return inf values

nonconvexopt committed:
nit: fix typo and specify dtype and device for tensor

nonconvexopt committed:
try using pinv instead of inv

nonconvexopt committed:
return inf for .mean with invalid parameters

nonconvexopt committed:
increase sample size in moment test and check convergence

nonconvexopt committed:
nit: remove trailing spaces

nonconvexopt committed:
fix shaping error

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution

nonconvexopt committed:
improve clarity for variable shapes in .variance

nonconvexopt committed:
match shape of variables in .mean

nonconvexopt committed:
strengthen moment test

nonconvexopt committed:
apply more negligable numerical stability care

nonconvexopt committed:
increase sample size in sample test proportional to the number of parameters

nonconvexopt committed:
fit flake8

nonconvexopt committed:
apply stronger numerical stability

nonconvexopt committed:
debug implementation error in .entropy

nonconvexopt committed:
Merge branch 'master' into inverse_wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution

nonconvexopt committed:
debug implementation error in .entropy

nonconvexopt committed:
debug implementation error in .entropy

nonconvexopt committed:
remove unnecessary warnings

nonconvexopt committed:
corrected implementation error in _bartlett_sampling

Commit:
Fix seed in distributions test for deterministic results (#76357)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76357

Fix for https://github.com/pytorch/pytorch/issues/76160

This fixes the random seed for the `test_wishart_log_prob` (like all random tests in `test_distributions`) to prevent non-determinism.

Test Plan: Tested locally.

Differential Revision: D35914795

fbshipit-source-id: 85001cc500e1d48d4e81c2e33b783ed6c38c9851

nonconvexopt committed:
Merge remote-tracking branch 'neerajprad/export-D35914795' into inverse_wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution

nonconvexopt committed:
expand df parameter test range

nonconvexopt committed:
Merge branch 'inverse_wishart_distribution' of github.com:nonconvexopt/pytorch into inverse_wishart_distribution

nonconvexopt committed:
modify test parameters to fit the purpose

nonconvexopt committed:
Merge branch 'master' into inverse_wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution

nonconvexopt committed:
Merge branch 'pytorch:master' into inverse_wishart_distribution
